crates/bevy_ecs/src/world/interior_mutable_world.rs

```diff
@@ -0,0 +1,498 @@
+///     let component_access = OnlyComponentAccessWorld(interior_mutable_world);
+///     (resource_access, component_access)
+/// }
+/// ```
+#[derive(Copy, Clone)]
+pub struct InteriorMutableWorld<'w>(&'w World);
+
+impl<'w> InteriorMutableWorld<'w> {
+    pub(crate) fn new(world: &'w World) -> Self {
+        InteriorMutableWorld(world)
+    }
+
+    /// Gets a reference to the [`&World`](crate::world::World) this [`InteriorMutableWorld`] belongs to.
+    /// This can be used to call methods like [`World::contains_resource`] which aren't exposed and but don't perform any accesses.
+    ///
+    /// **Note**: You *must not* hand out a `&World` reference to arbitrary safe code when the [`InteriorMutableWorld`] is currently
+    /// being used for mutable accesses.
+    ///
+    /// # Safety
+    /// - the world must not be used to access any resources or components. You can use it to safely access metadata.
+    pub unsafe fn world(&self) -> &'w World {
+        self.0
+    }
+
+    /// Retrieves this world's [Entities] collection
+    #[inline]
+    pub fn entities(&self) -> &'w Entities {
+        &self.0.entities
+    }
+
+    /// Retrieves this world's [Archetypes] collection
+    #[inline]
+    pub fn archetypes(&self) -> &'w Archetypes {
+        &self.0.archetypes
+    }
+
+    /// Retrieves this world's [Components] collection
+    #[inline]
+    pub fn components(&self) -> &'w Components {
+        &self.0.components
+    }
+
+    /// Retrieves this world's [Storages] collection
+    #[inline]
+    pub fn storages(&self) -> &'w Storages {
+        &self.0.storages
+    }
+
+    /// Retrieves this world's [Bundles] collection
+    #[inline]
+    pub fn bundles(&self) -> &'w Bundles {
+        &self.0.bundles
+    }
+
+    /// Reads the current change tick of this world.
+    #[inline]
+    pub fn read_change_tick(&self) -> u32 {
+        self.0.read_change_tick()
+    }
+
+    #[inline]
+    pub fn last_change_tick(&self) -> u32 {
+        self.0.last_change_tick()
+    }
+
+    #[inline]
+    pub fn increment_change_tick(&self) -> u32 {
+        self.0.increment_change_tick()
+    }
+
+    /// Retrieves an [`InteriorMutableEntityRef`] that exposes read and write operations for the given `entity`.
+    /// Similar to the [`InteriorMutableWorld`], you are in charge of making sure that no aliasing rules are violated.
+    pub fn get_entity(&self, entity: Entity) -> Option<InteriorMutableEntityRef<'w>> {
+        let location = self.0.entities.get(entity)?;
+        Some(InteriorMutableEntityRef {
+            world: InteriorMutableWorld(self.0),
+            entity,
+            location,
+        })
+    }
+
+    /// Gets a reference to the resource of the given type if it exists
+    ///
+    /// # Safety
+    /// It is the callers responsibility to ensure that
+    /// - the [`InteriorMutableWorld`] has permission to access the resource
+    /// - no other mutable references to the resource exist at the same time
```

Comment:
```suggestion
    /// - no mutable references to the resource exist at the same time
```

Same for the other read-only accesses.